js/bundle/src/encoder.ts

```diff
@@ -0,0 +1,141 @@
+        this.addIndexEntry(url, this.addResponse(new HeaderMap(status, headers), payload));
+    }
+
+    setManifestURL(url: string) {
+        validateExchangeURL(url);
+        this.addSection('manifest', url);
+    }
+
+    private addSection(name: string, content: CBORValue) {
+        if (this.sectionLengths.includes(name)) {
+            throw new Error('Duplicated section: ' + name);
+        }
+        this.sectionLengths.push(name);
+        this.sectionLengths.push(encodedLength(content));
+        this.sections.push(content);
+    }
+
+    private addResponse(headerMap: HeaderMap, payload: Uint8Array): number {
+        if (payload.length > 0 && !headerMap.has('content-type')) {
+            throw new Error('Non-empty exchange must have Content-Type header');
+        }
+
+        let response = [new Uint8Array(CBOR.encodeCanonical(headerMap)), payload];
+        this.responses.push(response);
+        return encodedLength(response);
+    }
+
+    private addIndexEntry(url: string, responseLength: number) {
+        this.index[url] = [
+            new Uint8Array(0), // variants-value
+            this.currentResponsesOffset,
+            responseLength,
+        ];
+        this.currentResponsesOffset += responseLength;
+    }
+
+    private fixupIndex() {
+        // Adjust the offsets by the length of the response section's CBOR header.
+        let responsesHeaderSize = encodedLength(this.responses.length);
+        for (let key in this.index) {
+            this.index[key][1] += responsesHeaderSize;
+        }
+        return this.index;
+    }
+
+    private createTopLevel(): CBORValue {
+        return [
+            byteString('🌐📦'),
+            byteString('b1\0\0'),
+            this.primaryURL,
+            new Uint8Array(CBOR.encodeCanonical(this.sectionLengths)),
+            this.sections,
+            new Uint8Array(8),  // Length (to be filled in later)
+        ];
+    }
+};
+
+class HeaderMap extends Map<string, string> {
+    constructor(status: number, headers: {[key:string]: string}) {
+        super();
+        if (status < 100 || status > 999) {
+            throw new Error('Invalid status code');
+        }
+
+        this.set(':status', status.toString());
+        for (let key in headers) {
+            this.set(key.toLowerCase(), headers[key]);
+        }
+    }
+
+    // This tells the CBOR library how to serialize this object.
+    encodeCBOR(encoder: CBOR.Encoder) {
+        // Convert keys and values to Uint8Array, as the CBOR representation of
+        // header map is {bytestring => bytestring}.
+        let m = new Map<Uint8Array, Uint8Array>();
+        for (let [key, value] of this.entries()) {
+            m.set(byteString(key), byteString(value));
+        }
+        return encoder.pushAny(m);
+    }
+}
+
+// Throws an error if `urlString` is not a valid exchange URL.
+function validateExchangeURL(urlString: string): void {
+    let url = new URL(urlString);
```

Comment:
This could throw if urlString is a relative URL, right?  Maybe write about that somewhere?

Should we also error out if protocol is not http/https?

Reply:
Done.